### ISSUE TITLE
Identical Or patterns

### ISSUE THREAD

**Issue by ndmitchell** [rust-lang/rust-clippy#11026]
### What it does

In pattern matching, if you write `Pattern | Pattern => ...` then it should suggest writing `Pattern => ...`. Duplicating a pattern twice in an Or pattern is unnecessary.

### Advantage

- Reduces the amount of code.
- In many cases will spot bugs, when the code was meant to be different.

### Drawbacks

None I can think of.

### Example

```rust
match x {
    (Ty::Tuple(_), _) | (Ty::Tuple(_), _) => true,
    _ => false,
}
```

Could be written as:

```rust
match x {
    (Ty::Tuple(_), _) => true,
    _ => false,
}
```

**Comment by Centri3** [comment]
Hmm, is there any reason why the rustc lint `unreachable_patterns` doesn't already work? It isn't specialized in saying "this pattern is the same as the previous", sure, but is there any reason that wouldn't work

**Comment by ndmitchell** [comment]
It seems this could equally be a warning in the compiler.

**Comment by workingjubilee** [comment]
It is fairly common to write code that looks like this:
```rust
match something {
    CONSTANT | RANGE_START..=RANGE_END => todo!(),
    _ => todo()
}
```
And for the constant to be within `RANGE_START..=RANGE_END`. This is often because the "constant" is a compile-time variable based on something like architecture, and also just because it can be nicer to read.
